src/S3Writer.php

```diff
@@ -0,0 +1,106 @@
+    public function __construct(Config $config, Logger $logger)
+    {
+        $this->config = $config;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Runs data extraction
+     * @throws \Exception
+     */
+    public function execute(string $sourcePath): void
+    {
+        try {
+            $client = new S3MultiRegionClient([
+                'version' => '2006-03-01',
+                'credentials' => [
+                    'key' => $this->config->getAccessKeyId(),
+                    'secret' => $this->config->getSecretAccessKey(),
+                ],
+            ]);
+            $region = $client->getBucketLocation(["Bucket" => $this->config->getBucket()])->get('LocationConstraint');
+            $client = new S3Client([
+                'region' => $region,
+                'version' => '2006-03-01',
+                'credentials' => [
+                    'key' => $this->config->getAccessKeyId(),
+                    'secret' => $this->config->getSecretAccessKey(),
+                ],
+            ]);
+
+            $finder = (new Finder())->in($sourcePath)->files();
+            /** @var SplFileInfo $file */
+            foreach ($finder as $file) {
+                // Remove initial forwardslash
+                $prefix = $this->config->getPrefix();
+                if (substr($prefix, 0, 1) == '/') {
+                    $prefix = substr($prefix, 1);
+                }
+                $key = $prefix . $file->getRelativePathname();
+                $this->logger->info("Uploading file {$file->getRelativePathname()} to {$key}");
+                $client->putObject(
+                    [
+                        'Bucket' => $this->config->getBucket(),
+                        'Key' => $key,
+                        'Body' => fopen($file->getPathname(), 'r'),
+                    ]
+                );
+            }
+        } catch (S3Exception $e) {
+            if ($e->getStatusCode() === 403) {
+                throw new UserException("Invalid credentials or permissions.", $e->getCode(), $e);
+            }
+            if ($e->getStatusCode() === 400 || $e->getStatusCode() === 401 || $e->getStatusCode() === 404) {
+                if (get_class($e->getPrevious()) === ClientException::class) {
+                    /** @var ClientException $previous */
+                    $previous = $e->getPrevious();
+                    if ($previous->getResponse()) {
+                        throw new UserException(
+                            $previous->getResponse()->getStatusCode()
+                            . " "
+                            . $previous->getResponse()->getReasonPhrase()
+                            . " ("
+                            . $e->getAwsErrorCode()
+                            . ")\n"
+                            . $previous->getResponse()->getBody()->__toString()
+                        );
+                    } else {
+                        throw new UserException($previous->getMessage());
```

Comment:
Tohle může být přímo, nepotřebuje to else, ne?

Reply:
👍